### PR TITLE
Improved upload/download methods for python client

### PR DESCRIPTION
**Pull request by bw27492**
- Progress bars for uploads and downloading to disk
- Method to download _all_ files from a release
- Ability to filter files based on regex (e.g. "*.json")

### CODE REVIEW

File: lib/python/src/bailo/helper/release.py [hunk 1]
```diff
@@ -171,12 +230,26 @@ def upload(self, path: str, data: BytesIO | None = None) -> str:
...
+            colour = "white"
         else:
-            res = self.client.simple_upload(self.model_id, name, data).json()
+            colour = "blue"
+
+        with tqdm(
+            total=size, unit="B", unit_scale=True, unit_divisor=1024, postfix=f"uploading {name}", colour=colour
+        ) as t:
+            wrapped_buffer = CallbackIOWrapper(t.update, data, "read")
+            res = self.client.simple_upload(self.model_id, name, wrapped_buffer).json()
```
Comment (GB27247) [inline lib/python/src/bailo/helper/release.py:252]: The progress bar seems to complete even if the request fails. I tested this with uploading two of the same file name.

File: lib/python/src/bailo/helper/release.py [hunk 2]
```diff
@@ -148,11 +157,61 @@ def download(self, filename: str, write: bool = True, path: str | None = None) -
...
+                colour=colour,
+            ) as t:
+                with open(path, "wb") as f:
+                    for data in res.iter_content(block_size):
+                        t.update(len(data))
+                        f.write(data)
 
         return res
 
+    def download_all(self, path: str = os.getcwd(), include: Union[list, str] = None, exclude: Union[list, str] = None):
```
Comment (GB27247) [inline lib/python/src/bailo/helper/release.py:183]: Old an new type annotations are being used here instead we should use `list | str | None`

**Review by GB27247 — CHANGES_REQUESTED** [review]

File: lib/python/src/bailo/helper/release.py [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
+from tqdm.utils import CallbackIOWrapper
 
 from bailo.core.client import Client
+from bailo.core.exceptions import BailoException
 from semantic_version import Version
 
+try:
+    no_color = int(os.environ["NO_COLOR"])
+except KeyError:
+    no_color = 0
```
[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py:18]: Done

[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py:18]: Done

[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py]: done

[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py]: done

File: lib/python/tests/test_files.py [hunk 4]
```diff
@@ -16,12 +17,39 @@ def test_file_upload(example_model):
...
+    for filename in ["test.json", "test2.txt", "to_exclude.txt"]:
+        example_release.upload(filename, file)
+        file.seek(0)
+
+    example_release.download_all(path=downloads_path, include=["*.txt"], exclude=["to_exclude.txt"])
+
+    assert os.listdir(downloads_path) == ["test2.txt"]
+
+    with open(str(downloads_path.join("test2.txt")), "rb") as f:
+        assert f.read() == byte_obj
```
[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/tests/test_files.py:42]: Done (split tests). Note - using the tmpdir fixture so writing to files isn't an issue

File: lib/python/src/bailo/helper/release.py [hunk 5]
```diff
@@ -148,11 +157,61 @@ def download(self, filename: str, write: bool = True, path: str | None = None) -
...
+
+        if exclude is not None:
+            file_names = [
+                file for file in file_names if not any([fnmatch.fnmatch(file, pattern) for pattern in exclude])
+            ]
+
+        os.makedirs(path, exist_ok=True)
+        for file in file_names:
+            file_path = os.path.join(path, file)
+            self.download(filename=file, path=file_path)
```
[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py:209]: Not necessary - `os.makedirs` provides permission denied error

File: lib/python/src/bailo/helper/release.py [hunk 6]
```diff
@@ -148,11 +157,61 @@ def download(self, filename: str, write: bool = True, path: str | None = None) -
         if write:
             if path is None:
                 path = filename
-            with open(path, "wb") as f:
-                f.write(res.content)
+            total_size = int(res.headers.get("content-length", 0))
+            block_size = 1024
```
[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py:157]: Done

File: lib/python/src/bailo/helper/release.py [hunk 7]
```diff
@@ -148,11 +157,61 @@ def download(self, filename: str, write: bool = True, path: str | None = None) -
...
+    def download_all(self, path: str = os.getcwd(), include: Union[list, str] = None, exclude: Union[list, str] = None):
+        """Writes all files to disk given a local directory.
+
+        :param include: List of regex statements for file names to include, defaults to None
+        :param exclude: List of regex statements for file names to exclude, defaults to None
+        :param path: Local directory to write files to
+        :raises BailoException: If the release has no files assigned to it
+        ..note:: Regex statements must be complete to work as intended (uses re.match not re.search)
+        """
+        files_metadata = self.client.get_release(self.model_id, self.version)["release"]["files"]
```
[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py:188]: Done

File: lib/python/src/bailo/helper/release.py (hunk 1 above)
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py:248]: Done

File: lib/python/src/bailo/helper/release.py (hunk 2 above)
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py:179]: Done

**Review by GB27247 — CHANGES_REQUESTED** [review]

File: lib/python/src/bailo/core/utils.py [hunk 8]
```diff
@@ -1,6 +1,9 @@
 from __future__ import annotations
 
 from typing import Any
+import os
+
+NO_COLOR = int(os.environ.get("NO_COLOR", 0))
```
[the remark this answers is not in the document]
Comment (bw27492) [reply lib/python/src/bailo/core/utils.py]: Done

File: lib/python/src/bailo/core/utils.py [hunk 9]
```diff
@@ -3,7 +3,7 @@
 from typing import Any
 import os
 
-NO_COLOR = int(os.environ.get("NO_COLOR", 0))
+NO_COLOR = bool(int(os.environ.get("NO_COLOR", 0)))
```
Comment (GB27247) [inline lib/python/src/bailo/core/utils.py:6]: This will Falsify strings such as "0". I would instead use something like, `"NO_COLOR" in os.environ`

**Review by GB27247 — CHANGES_REQUESTED** [review]

**Review by GB27247 — APPROVED** [review]

File: lib/python/src/bailo/helper/release.py [hunk 10]
```diff
@@ -171,12 +227,26 @@ def upload(self, path: str, data: BytesIO | None = None) -> str:
                 name = path
 
             with open(path, "rb") as f:
-                res = self.client.simple_upload(self.model_id, name, f).json()
+                data = BytesIO(f.read())
```
Comment (a3957273) [inline lib/python/src/bailo/helper/release.py:230]: This won't work with large files, it loads the entire file contents into memory.

**Review by a3957273 — CHANGES_REQUESTED** [review]

File: lib/python/src/bailo/helper/release.py (hunk 10 above)
Comment (bw27492) [reply lib/python/src/bailo/helper/release.py]: Done. Added new integration test to prevent this in future

**Review by a3957273 — APPROVED** [review]

**Pull request merged**